### DISCUSSION TITLE
CasePaths subscript not working on Swift 6

### DISCUSSION THREAD

**Discussion by gianfilippocisternino** [pointfreeco/swift-case-paths#203] [General]
Compiler is giving unexpected errors when using `CasePath` subscript on Swift 6. This error happens only in some cases (see snippet below).

While looking for a potential bug i found this commented code: https://github.com/pointfreeco/swift-case-paths/commit/e39069d748d14ec7c050198ad8659e18f64e0799

May that fix the issue?

This snippet compiles fine:
```swift
let index = state.navigationPath[topMainDestination]?.lastIndex(
  where: { destination in
    guard let modalDestination = destination[case: \.modal] as? MD else {
      return false
    }

    return modalDestination[case: keyPath] != nil
  }
)

guard
  let index,
  let destination = state.navigationPath[topMainDestination]?[index],
  let modalDestination = destination[case: \.modal] as? MD
else {
  return nil
}

return modalDestination[case: keyPath]
```

This one triggers an error somewhere else in the function:
```swift
return state.navigationPath[topMainDestination]?.compactMap { destination in
  guard let modalDestination = destination[case: \.modal] as? MD else {
    return nil
  }
  
  return modalDestination[case: keyPath]
}.last
```

**Example**
[CasePathsDemo.zip](https://github.com/user-attachments/files/17191453/CasePathsDemo.zip)

**Environment**
  - swift-case-paths 1.5.5
  - Xcode 16.0 (16A242)
  - Swift 6.0 (swiftlang-6.0.0.9.10 clang-1600.0.26.2)
  - OS: iOS 18.0 macOS 15.0 (24A335)

**Comment by mbrandonw** [comment]
Hi @gianfilippocisternino, can you please provide a minimal reproduction of the problem?

**Comment by gianfilippocisternino** [comment]
Hi @mbrandonw, i uploaded a sample app that shows the problem. Thanks

**Comment by mbrandonw** [comment]
Hi @gianfilippocisternino, this just seems like a regression in the Swift compiler in Swift 6. If you add some explicit type annotations it works just fine:

```swift
get: { [unowned self] () -> Value? in
…
return navigationPath[topMainDestination]?.compactMap { destination -> Value? in
```

Since this isn't an issue with the library I am going to convert it to a discussion. Please feel free to continue the conversation over there!
